relay-server/src/services/buffer/envelope_buffer/mod.rs

```diff
@@ -214,88 +246,128 @@ where
             QueueItem {
-                key: stack_key,
-                value: self.stack_provider.create_stack(envelope),
+                key: envelope_project_keys,
+                value: stack,
             },
             Priority::new(received_at),
         );
         debug_assert!(previous_entry.is_none());
-        for project_key in stack_key.iter() {
+        for project_key in envelope_project_keys.iter() {
             self.stacks_by_project
                 .entry(project_key)
                 .or_default()
-                .insert(stack_key);
+                .insert(envelope_project_keys);
         }
         relay_statsd::metric!(
             gauge(RelayGauges::BufferStackCount) = self.priority_queue.len() as u64
         );
+
+        Ok(())
     }
 
     /// Returns `true` if the underlying storage has the capacity to store more envelopes.
     pub fn has_capacity(&self) -> bool {
         self.stack_provider.has_store_capacity()
     }
 
-    fn pop_stack(&mut self, stack_key: StackKey) {
-        for project_key in stack_key.iter() {
+    /// Pops an [`EnvelopeStack`] with the supplied [`EnvelopeBufferError`].
+    fn pop_stack(&mut self, envelope_project_keys: EnvelopeProjectKeys) {
+        for project_key in envelope_project_keys.iter() {
             self.stacks_by_project
                 .get_mut(&project_key)
                 .expect("project_key is missing from lookup")
-                .remove(&stack_key);
+                .remove(&envelope_project_keys);
         }
-        self.priority_queue.remove(&stack_key);
+        self.priority_queue.remove(&envelope_project_keys);
 
         relay_statsd::metric!(
             gauge(RelayGauges::BufferStackCount) = self.priority_queue.len() as u64
         );
     }
+
+    /// Creates all the [`EnvelopeStack`]s with no data given a set of [`EnvelopeProjectKeys`].
+    async fn load_stacks(&mut self, envelopes_project_keys: HashSet<EnvelopeProjectKeys>) {
+        let received_at = Instant::now();
+
+        for envelope_project_keys in envelopes_project_keys {
+            let previous_entry = self.priority_queue.push(
+                QueueItem {
+                    key: envelope_project_keys,
+                    value: self.stack_provider.create_stack(envelope_project_keys),
+                },
+                // We don't set the readiness of the projects, because we will eventually poll
+                // the envelopes for this project keys pair and load the projects if not in cache,
+                // which, in turn, will mark the projects as ready via the `mark_ready` method.
+                Priority::new(received_at),
+            );
+
+            debug_assert!(previous_entry.is_none());
+
+            for project_key in envelope_project_keys.iter() {
+                self.stacks_by_project
+                    .entry(project_key)
+                    .or_default()
+                    .insert(envelope_project_keys);
+            }
+        }
+    }
```

Comment:
We should call `push_stack` here. That is, change the signature of `push_stack` to require a key pair instead of an envelope, and then reuse the function.

Reply:
I considered this, will try to generalize it further.